[SYCL][Doc] Update sycl_ext_oneapi_sub_group_mask (#8174)
intel/llvm: sycl/doc/extensions/supported/sycl_ext_oneapi_sub_group_mask.asciidoc

// docbook uses c++ and html5 uses cpp.
:language: {basebackend@docbook:c++:cpp}

== Notice

[%hardbreaks]

> Copyright (C) 2023-2023 Intel Corporation.  All rights reserved.

Reviewer: ```suggestion
Copyright (C) 2021-2023 Intel Corporation.  All rights reserved.
```

Author: Applied in 366c3ea5f81dbaec0d4429f593f6e74b72eec266

Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
permission by Khronos.

== Contact

To report problems with this extension, please open a new issue at:

https://github.com/intel/llvm/issues

== Dependencies

This extension is written against the SYCL 2020 revision 6 specification.  All
references below to the "core SYCL specification" or to section numbers in the
SYCL specification refer to that revision.

== Status

> This is a proposed extension specification, intended to gather community

Author: @Pennycook, could you please take a look at this change specifically? This is a "supported" extension, but status only says that it is "proposed". I used this status to better match with the previous version, which has several comments that the spec is only a draft.

I'm not entirely sure which exact status is the right one here.

Reviewer 2: Given that this is in the "supported" folder, I think the status is supposed to be:

> This extension is implemented and fully supported by DPC++.

@gmlueck - I had a quick look through the "supported" folder, and there are quite a few extensions that still have the old "Working Draft" wording (e.g. [sycl_ext_buffer_location](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_intel_buffer_location.asciidoc), [sycl_ext_intel_dataflow_pipes](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_intel_dataflow_pipes.asciidoc)).  Do we need somebody to go through all these extensions and update the status?

Author: Switched the status to "implemented" in 1aaaeca41d33dd1602bddaea3b17d82594d9b00c

|Function|Description

> |`template <typename Group> Group::mask_type group_ballot(Group g, bool predicate = true)`

Reviewer 2: ```suggestion
|`template <typename Group> sub_group_mask group_ballot(Group g, bool predicate = true)`
```

|Return a `sub_group_mask` with one bit for each work-item in group _g_. A bit
is set in this mask if and only if the corresponding work-item's _predicate_ is

> `true`.

Reviewer 2: ```suggestion
`true`. Available only when `std::is_same_v<std::decay_t<Group>, sub_group>` is true.
```

|Function|Description

> |`template <typename Group> Group::mask_type group_ballot(Group g, bool predicate = true)`

Reviewer 2: This is a hangover from a previous draft, where `group_ballot` supported other group types.  I think we should change the prototype to return `sub_group_mask` as you suggest.

> |`template <typename Group> Group::mask_type group_ballot(Group g, bool predicate = true)`

Reviewer 3: Since you asked for a re-review ... this issue is still not resolved.

> |`template <typename Group> Group::mask_type group_ballot(Group g, bool predicate = true)`

Author: This should be now resolved with 5ba66d3b52620bb71fa1d661ecddc41ba68c28bb

the work-item with id 0, and the most significant bit (MSB) corresponds to the
work-item with the id `max_local_range()-1`.

|===
|Constructor|Description

|`sub_group_mask()`
|Constructs a group mask with all bits set to 0. Size of a group mask

> corresponds to max local range of a sub-group which work-item belongs to.

Reviewer 2: ```suggestion
corresponds to max local range of the sub-group which work-item belongs to.
```

A work-item can only ever be a member of one sub-group.

Author: Applied in 366c3ea5f81dbaec0d4429f593f6e74b72eec266

|`sub_group_mask(unsigned long long val)`
|Constructs a group mask with the first `N` bit positions to the
corresponding bit values in _val_. `N` is a size of a group mask and it

> corresponds to max local range of a sub-group which work-item belongs to. If

Reviewer 2: ```suggestion
corresponds to max local range of the sub-group which work-item belongs to. If
```

Author: Applied in 366c3ea5f81dbaec0d4429f593f6e74b72eec266

> size of a group mask is bigger than number of bits in the value representation

Reviewer: ```suggestion
size of a group mask is bigger than the number of bits in the value representation
```

Author: Applied in 366c3ea5f81dbaec0d4429f593f6e74b72eec266

of `unsigned long long`, the remaining positions are initialized to zero.

> |`template <typename T> sub_group_mask(const T& &val)`
> |Constructs a group mask with the first `N` bit positions to the
> corresponding bit values in _val_. `T` must be a SYCL `marray` of integral
> types. `N` is a size of a group mask and it corresponds to max local range of a

Reviewer: > `T` must be a SYCL `marray` of integral types. 
```suggestion
|`template <typename T, std::size_t N> sub_group_mask(const sycl::marray<T, N> &val)`
|Constructs a group mask with the first `N` bit positions to the
corresponding bit values in _val_. `T` must be an integral type.
`N` is a size of a group mask and it corresponds to max local range of a
```

Author: Applied in 366c3ea5f81dbaec0d4429f593f6e74b72eec266, but with a slight change: `N` in the description is not the same as number of elements in `val`, but it is a max size of the sub-group current work-item belongs to.

> types. `N` is a size of a group mask and it corresponds to max local range of a

Reviewer 2: ```suggestion
types. `N` is a size of a group mask and it corresponds to max local range of the
```

Author: Applied in 366c3ea5f81dbaec0d4429f593f6e74b72eec266

the work-item with id 0, and the most significant bit (MSB) corresponds to the
work-item with the id `max_local_range()-1`.

|===

> |Constructor|Description

Reviewer 3: We should say somehow that all of these constructors and the assignment operator are available starting in revision 2 of this specification.

Author: Added corresponding note in 14c468290d540162908c8331758ab1db9cc57324

|`sub_group_mask()`
|Constructs a group mask with all bits set to 0. Size of a group mask
corresponds to max local range of the sub-group which work-item belongs to.

|`sub_group_mask(unsigned long long val)`
|Constructs a group mask with the first `N` bit positions to the
corresponding bit values in _val_. `N` is a size of a group mask and it
corresponds to max local range of the sub-group which work-item belongs to. If
size of a group mask is bigger than the number of bits in the value
representation of `unsigned long long`, the remaining positions are initialized
to zero.

|`template <typename T, std::size_t K> sub_group_mask(const sycl::marray<T, K>& &val)`
|Constructs a group mask with the first `N` bit positions to the
corresponding bit values in _val_. `T` must be a SYCL `marray` of integral
types. `N` is a size of a group mask and it corresponds to max local range of
the sub-group which work-item belongs to. If size of a group mask is bigger than
number of bits in the value representation of `T`, the remaining positions are
initialized to zero.

|`sub_group_mask(const sub_group_mask &other) = default`

> |Constructs a group mask as a copy of _other_.

Reviewer 3: Does this need similar language to what you have below about the sizes being the same?

Author: Yes, I think so. Added in 6b97b5b623e8a1236b48eac5631f658034edf1ba

|`sub_group_mask& operator=(const sub_group_mask &other) = default`
|Assigns this instance of `sub_group_mask` with a copy of _other_. Size of both

> group masks must be the same or otherwise behavior is undefined.

Reviewer 3: The user has no control over the size of a group mask because the size depends on the number of work-items in the sub-group, and this is implementation-defined.  Should we instead say that `other` must be created for the same sub-group as `this`?

Author: User can control sub-group size through `reqd_sub_group_size` attribute or through corresponding property. Also, masks are tied to _maximum_ sub-group size, meaning that size of a mask will be the same for all sub-groups in a kernel, i.e. we can "share" masks between work-items of a kernel just fine.

I'm fine with either variant, but I slightly prefer the existing wording: it describes the actual (kind of) limitation we have and thus provides users with freedom of doing various stuff. At the same time, I don't quite imagine a use case where `sub_group_mask` would be shared/transferred between work-items and therefore perhaps it is ok to "lock" it to a single sub-group.

@Pennycook, any thoughts here?

Reviewer 3: > User can control sub-group size through reqd_sub_group_size attribute or through corresponding property. Also, masks are tied to maximum sub-group size, meaning that size of a mask will be the same for all sub-groups in a kernel

Good point.  I was thinking that the size of the group mask was the number of work-items in the sub-group, but this is not the case.  I see now that it is the *maximum* number of work-items in a sub-group, and this is controllable by the user via `[[reqd_sub_group_size]]`.  I agree that the current wording is fine.

Reviewer 2: > @Pennycook, any thoughts here?

The only copying of sub-group masks I think we need to forbid is between kernels.  That covers the cases where two kernels have different `[[reqd_sub_group_size]]` attributes, and where two kernels are submitted to different devices.

I also agree that the current wording is fine.

the work-item with id 0, and the most significant bit (MSB) corresponds to the
work-item with the id `max_local_range()-1`.

|===
|Constructor|Description

|`sub_group_mask()`
|Constructs a group mask with all bits set to 0. Size of a group mask
corresponds to max local range of a sub-group which work-item belongs to.

|`sub_group_mask(unsigned long long val)`
|Constructs a group mask with the first `N` bit positions to the
corresponding bit values in _val_. `N` is a size of a group mask and it
corresponds to max local range of a sub-group which work-item belongs to. If
size of a group mask is bigger than number of bits in the value representation
of `unsigned long long`, the remaining positions are initialized to zero.

> |`template <typename T> sub_group_mask(const T& &val)`

Reviewer 2: I think @bader's suggestion fixes this, but I want to call it out just to be sure.  As written, would the constructor be ambiguous for `unsigned long long`?

Author: My assumption was that `marray` constructor would be implemented using some `enable_if` machinery, which would prevent ambiguity, but I guess without knowing the `N` template argument of `marray`, that wouldn't be possible

Author: Anyway, should be resolved by 366c3ea5f81dbaec0d4429f593f6e74b72eec266

  static constexpr size_t max_bits = /* implementation-defined */;

>   sub_group_mask();

Reviewer 3: It would be nice to add a comment here too saying that these are available starting in revision 2.

Author: Added corresponding note in 14c468290d540162908c8331758ab1db9cc57324
